sdk/state/update.go

```diff
@@ -0,0 +1,143 @@
+package state
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/stellar/experimental-payment-channels/sdk/txbuild"
+	"github.com/stellar/go/xdr"
+)
+
+type PaymentProposal struct {
+	IterationNumber            int64
+	ObservationPeriodTime      time.Duration
+	ObservationPeriodLedgerGap int64
+	AmountToInitiator          int64
+	AmountToResponder          int64
+	CloseSignatures            []xdr.DecoratedSignature
+	DeclarationSignatures      []xdr.DecoratedSignature
+}
+
+// TODO - validate inputs? (eg. no negative amounts)
+// TODO - payments to be in Amount struct
+// initiator will only call this
+func (c *Channel) NewPaymentProposal(payToInitiator int64, payToResponder int64) (*PaymentProposal, error) {
+	// TODO - remove
+	fmt.Println(c.Balance, payToResponder, payToInitiator)
+	newBalance := c.Balance + payToResponder - payToInitiator
+	amountToInitiator := int64(0)
+	amountToResponder := int64(0)
+	if newBalance > 0 {
+		amountToResponder = newBalance
+	} else {
+		amountToInitiator = newBalance * -1
+	}
+	txC, err := txbuild.Close(txbuild.CloseParams{
+		ObservationPeriodTime:      c.observationPeriodTime,
+		ObservationPeriodLedgerGap: c.observationPeriodLedgerGap,
+		InitiatorSigner:            c.localSigner.FromAddress(),
+		ResponderSigner:            c.remoteSigner,
+		InitiatorEscrow:            c.localEscrowAccount.Address,
+		ResponderEscrow:            c.remoteEscrowAccount.Address,
+		StartSequence:              c.startingSequence,
+		IterationNumber:            c.iterationNumber,
+		AmountToInitiator:          amountToInitiator,
+		AmountToResponder:          amountToResponder,
+	})
+	if err != nil {
+		return nil, err
+	}
+	txC, err = txC.Sign(c.networkPassphrase, c.localSigner)
+	if err != nil {
+		return nil, err
+	}
+	c.Balance += amountToResponder
+	c.Balance -= amountToInitiator
+	return &PaymentProposal{
+		ObservationPeriodTime:      c.observationPeriodTime,
+		ObservationPeriodLedgerGap: c.observationPeriodLedgerGap,
+		AmountToInitiator:          amountToInitiator,
+		AmountToResponder:          amountToResponder,
+		CloseSignatures:            txC.Signatures(),
+	}, nil
+}
+
+// TODO - what do to when you don't have the full other's KP? - Change to FullAddress?
+func (c *Channel) ConfirmPayment(p *PaymentProposal) (*PaymentProposal, error) {
+	// TODO - is c.initiator good way of differentiating?
+	if !c.initiator {
```

Comment:
The initiator and responder can both make payments, so sometimes the initiator will need to sign the close and decl tx.

I think instead of this if, we should think about what is safe fort he participant to sign.

The participant is safe to sign the close as long as the close is valid.

The participant is safe to sign the declaration, as long as the remote participant has signed the close.

If the above rules are checked for, we can write the signing code to only occur in those cases. That's the methodology I used for the confirm close.

Reply:
ya  that is definitely a better way of thinking about it. Made it much more clear imo
https://github.com/stellar/experimental-payment-channels/pull/55/commits/c84ca0309abd42cd75d04d1f491b82f8c89f1bac